File: drug-discovery-ui/backend/app/api/models/retro.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field


class RetroConstraints(BaseModel):
    time: Optional[str] = Field(None, description="fast|standard|extended")
    cost: Optional[str] = Field(None, description="low|medium|high")
    safety: Optional[str] = Field(None, description="standard|cautious|stringent")


class RetroRequest(BaseModel):
    target: str
    constraints: RetroConstraints = RetroConstraints()
    starting: Optional[List[str]] = None
    routes: int = Field(3, ge=1, le=5)


class RetroStep(BaseModel):
    id: str
    label: str
    yieldPct: Optional[float] = None
    reagents: Optional[str] = None
    conditions: Optional[str] = None


class RetroRoute(BaseModel):
    id: str
    steps: List[RetroStep]
    overallYield: Optional[float] = None
    cost: Optional[str] = None
    time: Optional[str] = None
    safety: Optional[str] = None
    difficulty: Optional[str] = None
    hazards: Optional[List[str]] = None
    greenScore: Optional[float] = None
    references: Optional[List[str]] = None


class RetroMeta(BaseModel):
    rankings: Dict[str, List[str]] = {}
    criticalSteps: Dict[str, List[str]] = {}
    alternatives: Dict[str, List[str]] = {}


class RetroResponse(BaseModel):
    ok: bool
    routes: List[RetroRoute] = []
    meta: Optional[RetroMeta] = None
    error: Optional[str] = None

File: drug-discovery-ui/src/services/retro.ts
import type { RetroConstraints } from '../features/retrosynthesis/RetroInputForm'

export type RetroStep = { id: string; label: string; yieldPct?: number; reagents?: string; conditions?: string }
export type RetroRoute = { id: string; steps: RetroStep[]; overallYield?: number; cost?: string; time?: string; safety?: string; difficulty?: string; hazards?: string[]; greenScore?: number; references?: string[] }

export async function planRetrosynthesis(input: { target: string; constraints: RetroConstraints; starting?: string[]; routes?: number }): Promise<{ ok: boolean; routes: RetroRoute[]; meta?: any; error?: string }> {
  const resp = await fetch('/api/v1/retro/plan', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ...input, routes: input.routes ?? 3 }),
  })
  if (!resp.ok) throw new Error(`HTTP ${resp.status}`)
  return (await resp.json()) as any
}
